### PR TITLE
[MOD] : Migrer MerQure.Tools vers .net standard.

### PR DESCRIPTION
**Pull request by Yingqing**
- [x] Migrer MerQure.Tools vers .net standard.

- [ ] Correction des défaut de jeunesse.

### CODE REVIEW

File: samples/MerQure.Tools.Samples/RetryBusExample/Domain/ActionService.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            catch (Exception e)
+            {
+                _sampleService.SendOnError(sample);
+                //...
+            }
+        }
+
+        private async Task ManageNewSample(Sample sample)
+        {
+            _fakeError++;
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Domain/ActionService.cs]: CRITICAL Make the enclosing instance method 'static' or remove this set on the 'static' field. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)

File: samples/MerQure.Tools.Samples/RetryBusExample/Domain/ActionService.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+            int value = new Random().Next(1000, 3000);
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Domain/ActionService.cs]: CRITICAL Make sure that using this pseudorandom number generator is safe here. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2245)

File: samples/MerQure.Tools.Samples/RetryBusExample/Domain/ActionService.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            await Task.Delay(value);
+            if (_fakeError % 15 == 0)
+            {
+                throw new Exception("This is an unmanageable exception, like NPE");
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Domain/ActionService.cs]: MAJOR 'System.Exception' should not be thrown by user code. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS112)

File: src/MerQure.Tools/RetryBusService.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            RetryConsumer<T> consumer = new RetryConsumer<T>(_messagingService, configuration, producer);
+
+            return new Bus<T>(producer, consumer);
+        }
+
+        private void ApplyNewConfiguration(RetryStrategyConfiguration configuration)
+        {
+            if (configuration.Channels == null || !configuration.Channels.Any())
+            {
+                throw new ArgumentNullException(nameof(configuration.Channels));
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/RetryBusService.cs]: MAJOR The parameter name 'Channels' is not declared in the argument list. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3928)

File: src/MerQure.Tools/RetryBusService.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            }
+
+            if(configuration.DelaysInMsBetweenEachRetry == null)
+            {
+                throw new ArgumentNullException(nameof(configuration.Channels));
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/RetryBusService.cs]: MAJOR The parameter name 'Channels' is not declared in the argument list. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3928)

File: src/MerQure.Tools/Buses/Publisher.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            {
+                TryPublishWithBrokerAcknowledgement(publisher, bindingValue, JsonConvert.SerializeObject(encapsuledMessage));
+            }
+        }
+        
+        public void PublishOnRetryExchange(Channel channel, T message, RetryInformations retryInformations)
+        {
+            List<int> delays = _messageBrokerConfiguration.DelaysInMsBetweenEachRetry;
+            int delay = 0;
+            if (delays.Count() >= retryInformations.NumberOfRetry)
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/Publisher.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/MerQure.Tools/Buses/Publisher.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            using (var publisher = _messagingService.GetPublisher($"{_messageBrokerConfiguration.BusName}.{RetryStrategyConfiguration.ErrorExchangeSuffix}", true))
+            {
+                TryPublishWithBrokerAcknowledgement(publisher, errorChanel, JsonConvert.SerializeObject(retryMessage));
+            }
+        }
+
+        internal void TryPublishWithBrokerAcknowledgement(IPublisher publisher, string channelName, string message)
+        {
+            bool published = publisher.PublishWithAcknowledgement(channelName, message);
+            if (published == false)
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/Publisher.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1125)

File: samples/MerQure.Tools.Samples/Startup.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace MerQure.Tools.Samples
+{
+	public class Startup
+    {
+        public IConfiguration Configuration { get; set; }
+
+        public Startup()
+        {
+			var p = Directory.GetCurrentDirectory();
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/Startup.cs]: MINOR Remove this unused 'p' local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1481)

File: samples/MerQure.Tools.Samples/Program.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
+﻿using MerQure.Tools.Samples.RetryBusExample.Domain;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+
+namespace MerQure.Tools.Samples
+{
+	public class Program
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/Program.cs]: MAJOR Add a 'protected' constructor or the 'static' keyword to the class declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1118)

File: src/MerQure.Tools/Buses/RetryConsumer.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+using MerQure.Messages;
+using MerQure.Tools.Exceptions;
+
+namespace MerQure.Tools.Buses
+{
+    internal class RetryConsumer<T> : Consumer<T> where T : IDelivered
+    {
+        private readonly Publisher<T> _producer;
+
+        private RetryStrategyConfiguration _retryConfiguration;
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/RetryConsumer.cs]: MAJOR Make '_retryConfiguration' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/MerQure.Tools/Buses/RetryConsumer.cs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+
+        private bool IsGoingToErrorExchange(RetryInformations technicalInformations)
+        {
+            if (!_retryConfiguration.DelaysInMsBetweenEachRetry.Any())
+            {
+                return true;
+            }
+            if (_retryConfiguration.MessageIsGoingIntoErrorBusAfterAllRepeat
+                   && technicalInformations.NumberOfRetry == _retryConfiguration.DelaysInMsBetweenEachRetry.Count())
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/RetryConsumer.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/MerQure.Tools/Buses/RetryConsumer.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        public RetryConsumer(IMessagingService messagingService, RetryStrategyConfiguration retryConfiguration, Publisher<T> producer) : base(messagingService)
+        {
+            _retryConfiguration = retryConfiguration;
+            _producer = producer;
+        }
+
+        internal void SendToErrorExchange(Channel channel, T deliveredMessage)
+        {
+            if (String.IsNullOrEmpty(deliveredMessage.DeliveryTag))
+                throw new ArgumentNullException(nameof(deliveredMessage.DeliveryTag));
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/RetryConsumer.cs]: MAJOR The parameter name 'DeliveryTag' is not declared in the argument list. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3928)

File: src/MerQure.Tools/Buses/RetryConsumer.cs [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            {
+                _producer.PublishOnRetryExchange(channel, message, retryInformations);
+            }
+            return messageInformations;
+        }
+        
+        public MessageInformations ApplyRetryStrategy(Channel channel, T deliveredMessage)
+        {
+            if (String.IsNullOrEmpty(deliveredMessage.DeliveryTag))
+                throw new ArgumentNullException(nameof(deliveredMessage.DeliveryTag));
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/RetryConsumer.cs]: MAJOR The parameter name 'DeliveryTag' is not declared in the argument list. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3928)

File: src/MerQure.Tools/Buses/Consumer.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        public void OnMessageReceived(EventHandler<T> callback, IMessagingEvent messagingEvent)
+        {
+            RetryMessage<T> retryMessage = JsonConvert.DeserializeObject<RetryMessage<T>>(messagingEvent.Message.GetBody());
+            retryMessage.OriginalMessage.DeliveryTag = EncodeDeliveryTag(messagingEvent.DeliveryTag);
+            RetryInformations.Add(retryMessage.OriginalMessage.DeliveryTag, retryMessage.RetryInformations);
+
+            callback(this, retryMessage.OriginalMessage);
+        }
+
+        private static string EncodeDeliveryTag(string deliveryTag) //TODO CLEAN !! this is just a fast fix ...
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/Consumer.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/MerQure.Tools/Buses/Consumer.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            return $"{deliveryTag}_{Guid.NewGuid().ToString()}";
+        }
+
+        private static string DecodeDeliveryTag(string deliveryTag) //TODO CLEAN !! this is just a fast fix
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/Consumer.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/MerQure.Tools/Buses/ConsumerProvider.cs [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MerQure.Tools.Buses
+{
+    public class ConsumerProvider
+    {
+        private static object _syncRoot = new Object();
+
+        private Dictionary<string, IConsumer> _consumers = new Dictionary<string, IConsumer>();
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/ConsumerProvider.cs]: MAJOR Make '_consumers' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/MerQure.Tools/Exceptions/MerqureToolsException.cs [hunk 17]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+
+namespace MerQure.Tools.Exceptions
+{
+    public class MerqureToolsException : Exception
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Exceptions/MerqureToolsException.cs]: MAJOR Update this implementation of 'ISerializable' to conform to the recommended serialization pattern. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3925)

File: samples/MerQure.Tools.Samples/RetryBusExample/Domain/SampleException.cs [hunk 18]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MerQure.Tools.Samples.RetryBusExample.Domain
+{
+    public class SampleException : Exception
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Domain/SampleException.cs]: MAJOR Update this implementation of 'ISerializable' to conform to the recommended serialization pattern. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3925)

File: samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleService.cs [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+            SampleBus = _retryBusService.CreateNewBus<Sample>(sampleConfiguration);
+        }
+
+        public void Send(Sample sample)
+        {
+            SampleBus.Publish(SampleChannels.MessageSended, sample, true);
+        }
+
+        public void Consume(EventHandler<Sample> callback)
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleService.cs]: CRITICAL Rename parameter 'callback' to 'onSampleReceived' to match the interface declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS927)

File: samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleService.cs [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+using System;
+using System.Collections.Generic;
+
+
+namespace MerQure.Tools.Samples.RetryBusExample.RetryExchangeExample.Infra
+{
+    public class SampleService : ISampleService
+    {
+        public const string BusName = "sample";
+        private readonly IRetryBusService _retryBusService;
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleService.cs]: MINOR Remove the field '_retryBusService' and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1450)

File: src/MerQure.Tools/Buses/Bus.cs [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            Consumer.Consume(channel, callback);
+        }
+
+        public void Publish(Channel channel, T message, bool applyDeliveryDeplay = false)
+        {
+            Producer.Publish(channel, message, applyDeliveryDeplay);
+        }
+
+        public void PublishWithTransaction(Channel channel, IEnumerable<T> messages, bool applyDeliveryDeplay = false)
```
Comment (LuccaIntegration) [inline src/MerQure.Tools/Buses/Bus.cs]: CRITICAL Rename parameter 'messages' to 'message' to match the interface declaration. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS927)

File: samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleChannels.cs [hunk 22]
```diff
@@ -0,0 +1,21 @@
+﻿using MerQure.Tools.Configurations;
+using System.Collections.Generic;
+
+namespace MerQure.Tools.Samples.RetryBusExample.Infra
+{
+    public static class SampleChannels
+    { 
+        public static Channel MessageSended = new Channel("v1.sample.messagesended");
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleChannels.cs]: CRITICAL Change the visibility of 'MessageSended' or make it 'const' or 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2223)
MINOR Make this field 'private' and encapsulate it in a 'public' property. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1104)

File: samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleChannels.cs [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+
+        public static Channel OtherAction = new Channel("v1.sample.otheraction");
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleChannels.cs]: CRITICAL Change the visibility of 'OtherAction' or make it 'const' or 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2223)
MINOR Make this field 'private' and encapsulate it in a 'public' property. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1104)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 25 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Yingqing** [comment]
@ThibautRappet Oui. Set si tu veux, je te debrief un peu les changements. C pas très évident de voir sur la PR.

File: samples/MerQure.Tools.Samples/MerQure.Tools.Samples.csproj [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+    <PackageReference Include="Microsoft.Extensions.Options.ConfigurationExtensions" Version="2.1.1" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\MerQure.RbMQ\MerQure.RbMQ.csproj" />
+    <ProjectReference Include="..\..\src\MerQure.Tools\MerQure.Tools.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Compile Update="Startup.cs">
```
Comment (rducom) [inline samples/MerQure.Tools.Samples/MerQure.Tools.Samples.csproj]: inutile

File: samples/MerQure.Tools.Samples/MerQure.Tools.Samples.csproj [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+      <CopyToOutputDirectory>Never</CopyToOutputDirectory>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="rabbitMQ.json">
```
Comment (rducom) [inline samples/MerQure.Tools.Samples/MerQure.Tools.Samples.csproj]: là tu peux regrouper en *.json

File: tests/MerQure.Tools.Tests/MerQure.Tools.Tests.csproj [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Moq" Version="4.10.0" />
+    <PackageReference Include="xunit" Version="2.4.0" />
+    <PackageReference Include="xunit.analyzers" Version="0.10.0" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
```
Comment (rducom) [inline tests/MerQure.Tools.Tests/MerQure.Tools.Tests.csproj]: PrivateAssets et IncludeAssets inutiles

File: samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleService.cs [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+using System;
+using System.Collections.Generic;
+
+
+namespace MerQure.Tools.Samples.RetryBusExample.RetryExchangeExample.Infra
+{
+	public class SampleService : ISampleService
+    {
+        public const string BusName = "sample";
+        private readonly IRetryBusService _retryBusService;
```
Comment (LuccaIntegration) [inline samples/MerQure.Tools.Samples/RetryBusExample/Infra/SampleService.cs]: MINOR Remove the field '_retryBusService' and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1450)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 10 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**